src/v/raft/replicate_batcher.cc

```diff
@@ -25,93 +25,72 @@
-    return _lock
-      .with(
-        [this, r = std::move(r)]() mutable { return do_cache(std::move(r)); })
-      .then([this](item_ptr i) {
-          if (_pending_bytes >= _max_batch_size) {
-              _flush_timer.cancel();
-              dispatch_background_flush();
-              return i->_promise.get_future();
-          }
-
-          if (!_flush_timer.armed()) {
-              _flush_timer.rearm(clock_type::now() + _debounce_duration);
-          }
-          return i->_promise.get_future();
-      });
+    return do_cache(std::move(r)).then([this](item_ptr i) {
+        return _lock.with([this] { return flush(); }).then([i] {
+            return i->_promise.get_future();
+        });
+    });
 }
 
 ss::future<> replicate_batcher::stop() {
-    _flush_timer.cancel();
     // we keep a lock here to make sure that all inflight requests have finished
     // already
     return _lock.with([this]() {
-        if (_pending_bytes > 0) {
-            _ptr->_ctxlog.info(
-              "Setting exceptional futures to {} pending writes",
-              _item_cache.size());
-            for (auto& i : _item_cache) {
-                i->_promise.set_exception(std::runtime_error(
-                  "replicate_batcher destructor called. Cannot "
-                  "finish replicating pending entries"));
-            }
-            _item_cache.clear();
-            _data_cache.clear();
-            _pending_bytes = 0;
+        for (auto& i : _item_cache) {
+            i->_promise.set_exception(
+              std::runtime_error("replicate_batcher destructor called. Cannot "
+                                 "finish replicating pending entries"));
         }
+        _item_cache.clear();
+        _data_cache.clear();
     });
 }
 ss::future<replicate_batcher::item_ptr>
 replicate_batcher::do_cache(model::record_batch_reader&& r) {
     return model::consume_reader_to_memory(std::move(r), model::no_timeout)
       .then([this](ss::circular_buffer<model::record_batch> batches) {
-          auto i = ss::make_lw_shared<item>();
-          size_t record_count = 0;
-          for (auto& b : batches) {
-              record_count += b.record_count();
-              _pending_bytes += b.size_bytes();
-              if (b.header().ctx.owner_shard == ss::this_shard_id()) {
-                  _data_cache.emplace_back(std::move(b));
-              } else {
-                  _data_cache.emplace_back(b.copy());
-              }
-          }
-          i->record_count = record_count;
-          _item_cache.emplace_back(i);
-          return i;
+          ss::circular_buffer<model::record_batch> data;
+          size_t bytes = std::accumulate(
+            batches.cbegin(),
+            batches.cend(),
+            size_t{0},
+            [](size_t sum, const model::record_batch& b) {
+                return sum + b.size_bytes();
+            });
+          return ss::get_units(_max_batch_size_sem, bytes)
```

Comment:
I think you can move this to a different method to un-indent. 

`do_cache_with_backpressure( batches )`